File: tactics2d/math/interpolate/bezier.py
##! python3
# Copyright (C) 2024, Tactics2D Authors. Released under the GNU GPLv3.
# @File: bezier.py
# @Description: This file implements a Bezier curve interpolator.
# @Author: Yueyuan Li
# @Version: 1.0.0

import numpy as np
from cpp_function import Bezier as cpp_Bezier

class Bezier:
    """This class implement a Bezier curve interpolator.

    Attributes:
        order (int): The order of the Bezier curve. The order of a Bezier curve is equal to the number of control points minus one.
    """

    def __init__(self, order: int):
        """Initialize the Bezier curve interpolator.

        Args:
            order (int): The order of the Bezier curve.

        Raises:
            ValueError: The order of a Bezier curve must be greater than or equal to one.
        """
        self.order = order
        if self.order < 1:
            raise ValueError("The order of a Bezier curve must be greater than or equal to one.")

    def _check_validity(self, control_points: np.ndarray):
        if len(control_points.shape) != 2 or control_points.shape[1] != 2:
            raise ValueError("The shape of control_points is expected to be (n, 2).")

        if len(control_points) != self.order + 1:
            raise ValueError(
                "The number of control points must be equal to the order of the Bezier curve plus one."
            )

    def de_casteljau(self, points: np.ndarray, t: float, order: int) -> np.ndarray:
        """The de Casteljau algorithm for Bezier curves.

        Args:
            points (np.ndarray): The interpolation points of the curve. The shape is (order + 1, 2).
            t (float): The binomial coefficient.
            order (int): The order of the Bezier curve.
        """
        if order == 1:
            return points[0] * (1 - t) + points[1] * t
        else:
            new_points = points[:-1] * (1 - t) + points[1:] * t
            return self.de_casteljau(new_points, t, order - 1)

    def get_curve(self, control_points: np.ndarray, n_interpolation: int) -> np.ndarray:
        """
        Args:
            control_points (np.ndarray): The control points of the curve. The shape is (order + 1, 2).
            n_interpolation (int): The number of interpolations.

        Returns:
            curve_points (np.ndarray): The interpolated points of the curve. The shape is (n_interpolation, 2).
        """
        self._check_validity(control_points)

        # curve_points = []
        #
        # interpolates = np.linspace(control_points[:-1], control_points[1:], n_interpolation)
        # if self.order == 1:
        #     return np.squeeze(interpolates)
        #
        # for n in range(n_interpolation):
        #     t = n / (n_interpolation - 1)
        #     point = self.de_casteljau(interpolates[n], t, self.order - 1)
        #     curve_points.append(point)

        curve_points = cpp_Bezier.get_curve(control_points, n_interpolation)
        return np.array(curve_points)

File: tactics2d/math/cpp_function/src/bezier.cpp
#include "bezier.h"



std::vector<std::array<double, 2>> Bezier::get_curve(
    const std::vector<std::array<double, 2>>& control_points,
    int n_interpolation
) {
    /*
    Args:
        control_points (std::vector<std::array<double, 2>>): The control points of the curve. The shape is (order + 1, 2).
        n_interpolation (int): The number of interpolations.

    Returns:
        curve_points (std::vector<std::array<double, 2>>): The interpolated points of the curve. The shape is (n_interpolation, 2).
    */

    int N = control_points.size();
    std::vector<std::vector<double>> ta(N, std::vector<double>(N, 0.0));
    ta.reserve(N); // Reserve memory in advance
    // Initialize the left and right sides of Pascal's triangle to 1
    for (int i = 0; i < N; ++i) {
        ta[i][0] = 1;
        ta[i][i] = 1;
    }

     // Calculate Pascal's triangle
    for (int row = 2; row < N; ++row) {
        for (int col = 1; col < row; ++col) {
            ta[row][col] = ta[row - 1][col - 1] + ta[row - 1][col];
        }
    }

    std::vector<std::array<double, 2>> curve_points(n_interpolation, {0.0, 0.0});
    curve_points.reserve(n_interpolation); // Reserve memory in advance
    double delta_t = 1.0 / (n_interpolation - 1);
    double t = -delta_t;
    double one_minus_t = 1 - t;
    for (int i = 0; i < n_interpolation; ++i) {
        t += delta_t;
        one_minus_t -= delta_t;
        for (int k = 0; k < N; ++k) {
            int c = k;  
            int b = N - c - 1;  
            double a = ta[N - 1][k]; 

            double one_minus_t_pow_b = std::pow(one_minus_t, b);
            double t_pow_c = std::pow(t, c);
            double temp = a * one_minus_t_pow_b * t_pow_c;
            // Calculate the x and y coordinates of the point
            curve_points[i][0] += temp * control_points[k][0];
            curve_points[i][1] += temp * control_points[k][1];
        }
    }

    return curve_points;
}
